doc/pins/batch_requests_api_hld.md

```diff
@@ -0,0 +1,184 @@
+# Batch Requests API HLD
+
+## Table of Content
+
+* [Table of Content](#table-of-content)
+* [Revision](#revision)
+* [Scope](#scope)
+* [Definitions/Abbreviations](#definitions-abbreviations)
+* [Overview](#overview)
+* [Requirements](#requirements)
+* [Architecture Design](#architecture-design)
+* [High-Level Design](#high-level-design)
+  + [Batched APIs](#batched-apis)
+  + [Batched API Operations](#batched-api-operations)
+  + [Orchagent Update](#orchagent-update)
+  + [PINS P4RT](#pins-p4rt)
+* [SAI API](#sai-api)
+* [Configuration and management](#configuration-and-management)
+* [Warmboot and Fastboot Design Impact](#warmboot-and-fastboot-design-impact)
+* [Restrictions/Limitations](#restrictions-limitations)
+* [Testing Requirements/Design](#testing-requirements-design)
+  + [Unit Test cases](#unit-test-cases)
+* [Open/Action items - if any](#open-action-items---if-any)
+* [Appendix](#appendix)
+  + [Current ProducerStateTable APIs](#current-producerstatetable-apis)
+  + [Current ProducerStateTable Operations](#current-producerstatetable-operations)
+
+## Revision
+
+Rev  | Rev Date  | Author(s)                                   | Change Description
+---- | --------- | --------------------------------------------| ------------------
+v0.1 | 02/15/2022 | Runming Wu, Srikishen Pondicherry Shanmugam | Initial version
+
+## Scope
+
+This document describes the high-level design of the batch request API in APPL DB.
+
+## Definitions/Abbreviations
+
+**PINS**: P4 Integrated Network Stack (PINS) provides a remote interface to SAI using P4.
+
+## Overview
+
+When an application writes an entry into APPL DB, it uses the ProducerStateTable API. Currently, the ProducerStateTable API does not support batched writes. In the current form, the application needs to call the API multiple times for programming a batch of requests. This HLD proposes a new API in ProducerStateTable that can allow applications to program multiple entries in a single invocation. This can reduce the number of Redis DB notifications and potentially reduce the latency for responses when the response path is enabled.
```

Comment:
Could you explain "potentially reduce the latency for responses when the response path is enabled"?
1. "response path" is not well defined.
2. For latency issue, there is a pipeline feature developed. Please investigate if it fits your use case.

Reply:
1. Response path is referring to https://github.com/Azure/SONiC/pull/846. It is end-to-end sync mode from the application to the hardware. Should we add a reference here? The HLD pr is not merged yet.
2. The latency reduction in on orchagent making bulk SAI call. (For example, the p4orch route manager makes bulk SAI calls to program routes.) This feature will send one final notification at the end. Comparing to the individual calls, which will send notification for each call, this new batch API will prevent orchagent to wake up prematurely, so that orchagent can take full advantage of the batch in the bulk SAI call.